store-implementation/federated-store/src/main/java/uk/gov/gchq/gaffer/federatedstore/operation/handler/impl/FederatedOperationHandler.java

```diff
@@ -0,0 +1,114 @@
+import uk.gov.gchq.gaffer.federatedstore.util.FederatedStoreUtil;
+import uk.gov.gchq.gaffer.graph.Graph;
+import uk.gov.gchq.gaffer.operation.Operation;
+import uk.gov.gchq.gaffer.operation.OperationException;
+import uk.gov.gchq.gaffer.operation.io.Output;
+import uk.gov.gchq.gaffer.store.Context;
+import uk.gov.gchq.gaffer.store.Store;
+import uk.gov.gchq.gaffer.store.operation.handler.OperationHandler;
+import uk.gov.gchq.koryphe.Since;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.BiFunction;
+
+import static java.util.Objects.isNull;
+import static java.util.Objects.nonNull;
+
+/**
+ * FederatedOperation handler for the federation of an PAYLOAD operation with an expected return type OUTPUT
+ */
+@Since("2.0.0")
+public class FederatedOperationHandler<INPUT, OUTPUT> implements OperationHandler<FederatedOperation<INPUT, OUTPUT>> {
+
+    public static final String ERROR_WHILE_RUNNING_OPERATION_ON_GRAPHS = "Error while running operation on graphs";
+
+    @Override
+    public Object doOperation(final FederatedOperation<INPUT, OUTPUT> operation, final Context context, final Store store) throws OperationException {
+        final Iterable<?> allGraphResults = getAllGraphResults(operation, context, (FederatedStore) store);
+
+        return mergeResults(allGraphResults, operation, (FederatedStore) store);
+    }
+
+    private Iterable getAllGraphResults(final FederatedOperation<INPUT, OUTPUT> operation, final Context context, final FederatedStore store) throws OperationException {
+        try {
+            List<Object> results;
+            final Collection<Graph> graphs = getGraphs(operation, context, store);
+            results = new ArrayList<>(graphs.size());
+            for (final Graph graph : graphs) {
+
+                final Operation updatedOp = FederatedStoreUtil.updateOperationForGraph(operation.getUnClonedPayload(), graph);
+                if (updatedOp != null) {
+                    try {
+                        if (updatedOp instanceof Output) {
+                            results.add(graph.execute((Output) updatedOp, context));
+                        } else {
+                            graph.execute(updatedOp, context);
+                            if (nonNull(operation.getMergeFunction())) {
+                                results.add(null);
+                            }
+                        }
+                    } catch (final Exception e) {
+                        if (!operation.isSkipFailedFederatedExecution()) {
+                            throw new OperationException(FederatedStoreUtil.createOperationErrorMsg(operation, graph.getGraphId(), e), e);
+                        }
+                    }
+                }
+            }
+
+            return results;
+        } catch (
+                final Exception e) {
+            throw new OperationException(ERROR_WHILE_RUNNING_OPERATION_ON_GRAPHS, e);
```

Comment:
Isn't this only needed for line 61? It looks like that is the only line that can throw anything and isn't part of the other try/catch. As the message is `ERROR_WHILE_RUNNING_OPERATION_ON_GRAPHS` it looks like this is intended to cover only line 61.

Reply:
Its BIG catch for everything and anything (like dumb nullpointers) that will happen now or later to identify concisely where and what stage of federation something went wrong.


there are two parts "getting results from graphs" and "merging"